src/dtree.ml

```diff
@@ -0,0 +1,648 @@
+type action = (term_env, term) Bindlib.mbinder
+
+(** {b Example} Given a rewrite system for a symbol [f] given as
+    - [f Z (S m)     → S m]
+    - [f n Z         → n]
+    - [f (S n) (S m) → S (S m)]
+
+    A possible tree might be
+    {v
++–?–∘–Z–∘     → n
+├–Z–∘–Z–∘     → n
+|   └–S–∘–?–∘ → S m
+└–S–∘–Z–∘     → n
+    └–S–∘–?–∘ → S (S m)
+    v}
+    with [∘] being a node (with a label not shown here) and [–u–]
+    being an edge with a matching on symbol [u] or a variable or wildcard when
+    [?].  Typically, the portion [S–∘–Z] is made possible by a swap. *)
+
+(** {3 Reduction substrate} *)
+
+(** Element consumed when reducing terms.  When reducing, we must have
+    - fast access to any element in the substrate (for swaps)
+    - fast replacement of an element and extension to replace an element of
+      the substrate by its reduced form, or an unfolding of
+      {!constructor:Appl} nodes.
+
+    Arrays have fast access but can't be extended efficiently.  Lists can be
+    extended easily but access is slow.  Tree structure have moderately fast
+    access and extension, and are therefore appropriate.
+
+    Two implementations have been tried with
+    - {!module:BatVect},
+    - {!module:BatFingerTree}
+    The performances are similar. *)
+module type reduction_substrate =
+sig
+  (** Type of a substrate of ['a]. *)
+  type 'a t
+
+  (** The empty substrate. *)
+  val empty : 'a t
+
+  (** [is_empty v] returns whether a stack is empty. *)
+  val is_empty : 'a t -> bool
+
+  (** [of_list l] returns a stack containing the elements of [l]. *)
+  val of_list : 'a list -> 'a t
+
+  (** [length v] is the number of elements in [v]. *)
+  val length : 'a t -> int
+
+  (** Prefix and suffix of the substrate. *)
+  type 'a prefix
+  type 'a suffix
+
+  (** [destruct v i] returns a triplet [(r, e, o)] with [r] being the
+      elements from 0 to [i - 1], [e] the [i]th element and [o] the elements
+      from [i + 1] to the end of [v]. *)
+  val destruct : 'a t -> int -> 'a prefix * 'a * 'a suffix
+
+  (** [restruct r n o] is the concatenation of three stacks [r] [n] and
+      [o]. *)
+  val restruct : 'a prefix -> 'a list -> 'a suffix -> 'a t
+end
+
+module ReductionStack : reduction_substrate =
+struct
+  type 'a t = 'a list
+  type 'a prefix = 'a list
+  type 'a suffix = 'a list
+  let empty = []
+  let is_empty l = (=) [] l
+  let of_list l = l
+  let length = List.length
+  let destruct e i =
+    if i < 0 then invalid_arg "RedStack.destruct";
+    let rec destruct l i r =
+      match (r, i) with
+      | ([]  , _) -> raise Not_found
+      | (v::r, 0) -> (l, v, r)
+      | (v::r, i) -> destruct (v :: l) (i - 1) r
+    in
+    destruct [] i e
+  let restruct l c r =
+    let rec insert acc l =
+      match l with
+      | []   -> acc
+      | x::l -> insert (x :: acc) l
+    in
+    insert (c @ r) l
+end
+
+(** {3 Operators on trees} *)
+
+(** [iter l n f t] is a generic iterator on trees; with function [l] performed
```

Comment:
These functions on trees should be in basics.ml.

Reply:
I can't have functions depending on `Dtree` in basics, as I have functions depending on `Basics` in `Dtree`

Reply:
I don't understand. The type tree is defined in terms.ml. Basics.ml usually contains functions on types defined in terms.ml. Why iter, capacity and to_dot cannot be moved to basics.ml?

Reply:
Ah indeed I forgot that trees were defined in terms and not Dtree